File: ios/Runner/HdrImageView.swift
import Flutter
import UIKit

/// Displays a still image (HEIC HLG/PQ) with its HDR (EDR) luminance shown
/// as-is, via a plain `UIImageView`.
///
/// Flutter's own `Image` widget renders through a Skia texture, which
/// tone-maps HDR content down to SDR — a native view is required, same
/// reasoning as `HdrVideoPlayerView`.
class HdrImageViewFactory: NSObject, FlutterPlatformViewFactory {
  func create(
    withFrame frame: CGRect,
    viewIdentifier viewId: Int64,
    arguments args: Any?
  ) -> FlutterPlatformView {
    return HdrImageView(frame: frame, arguments: args)
  }

  func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
    return FlutterStandardMessageCodec.sharedInstance()
  }
}

class HdrImageView: NSObject, FlutterPlatformView {
  private let imageView: UIImageView

  init(frame: CGRect, arguments args: Any?) {
    imageView = UIImageView(frame: frame)
    imageView.contentMode = .scaleAspectFit
    super.init()

    guard let params = args as? [String: Any], let path = params["path"] as? String else {
      return
    }
    guard let image = UIImage(contentsOfFile: path) else { return }
    imageView.image = image
    if #available(iOS 17.0, *) {
      // Renders the HEIC's HLG/PQ luminance as EDR instead of tone-mapping
      // it down to SDR.
      imageView.preferredImageDynamicRange = .high
    }
  }

  func view() -> UIView {
    imageView
  }
}

File: android/app/src/main/kotlin/com/eonlineservice/hdr/HdrImageView.kt
package com.eonlineservice.hdr

import android.content.Context
import android.graphics.BitmapFactory
import android.view.View
import android.widget.ImageView
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

private const val VIEW_TYPE_ID = "com.eonlineservice.hdr/hdr_image_view"

/**
 * Displays a still image (Ultra HDR JPEG or a plain JPEG) with its HDR
 * gainmap luminance shown as-is, via a plain [ImageView].
 *
 * Flutter's own `Image` widget renders through Skia's texture pipeline,
 * which doesn't apply an embedded Ultra HDR gainmap — a native View is
 * required, same reasoning as [HdrVideoPlayerView]. The window is already
 * switched to ActivityInfo.COLOR_MODE_HDR in MainActivity.onCreate.
 */
class HdrImageViewFactory(private val messenger: BinaryMessenger) :
    PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        @Suppress("UNCHECKED_CAST")
        val creationParams = args as? Map<String, Any?>
        return HdrImageView(context, creationParams)
    }
}

class HdrImageView(context: Context, creationParams: Map<String, Any?>?) : PlatformView {
    private val imageView = ImageView(context).apply {
        scaleType = ImageView.ScaleType.FIT_CENTER
    }

    init {
        val path = creationParams?.get("path") as? String
        if (path != null) {
            // decodeFile preserves an embedded Ultra HDR gainmap (API 34+);
            // on older devices it just decodes the SDR base layer.
            imageView.setImageBitmap(BitmapFactory.decodeFile(path))
        }
    }

    override fun getView(): View = imageView

    override fun dispose() {}
}
